File: LLD/Design-Patterns/Command/code/command.py
"""
Command Pattern — a RemoteControl (Invoker) triggers Commands against a Light (Receiver), with undo support.
Run: python command.py
"""

from abc import ABC, abstractmethod


class Light:
    """Receiver — knows how to actually perform the operations."""

    def __init__(self):
        self.on = False

    def turn_on(self):
        self.on = True
        print("Light is ON")

    def turn_off(self):
        self.on = False
        print("Light is OFF")


class RemoteCommand(ABC):
    @abstractmethod
    def execute(self):
        ...

    @abstractmethod
    def undo(self):
        ...


class LightOnCommand(RemoteCommand):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.turn_on()

    def undo(self):
        self.light.turn_off()


class LightOffCommand(RemoteCommand):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.turn_off()

    def undo(self):
        self.light.turn_on()


class RemoteControl:
    """Invoker — triggers commands and keeps a history for undo."""

    def __init__(self):
        self.history = []

    def press_button(self, command: RemoteCommand):
        command.execute()
        self.history.append(command)

    def press_undo(self):
        if not self.history:
            print("Nothing to undo.")
            return
        last = self.history.pop()
        last.undo()


if __name__ == "__main__":
    light = Light()
    remote = RemoteControl()

    on_command = LightOnCommand(light)
    off_command = LightOffCommand(light)

    print("Press ON button:")
    remote.press_button(on_command)

    print("Press OFF button:")
    remote.press_button(off_command)

    print("Press UNDO button:")
    remote.press_undo()

    print("Press UNDO button again:")
    remote.press_undo()

    print("Press UNDO button once more (nothing left):")
    remote.press_undo()

File: LLD/Design-Patterns/Command/code/Command.java
// Command Pattern — a RemoteControl (Invoker) triggers Commands against a Light (Receiver), with undo support.
// Compile: javac Command.java
// Run:     java Command

import java.util.Stack;

// Receiver — knows how to actually perform the operations.
class Light {
    private boolean on = false;

    void turnOn() {
        on = true;
        System.out.println("Light is ON");
    }

    void turnOff() {
        on = false;
        System.out.println("Light is OFF");
    }
}

interface RemoteCommand {
    void execute();
    void undo();
}

class LightOnCommand implements RemoteCommand {
    private Light light;

    LightOnCommand(Light light) {
        this.light = light;
    }

    public void execute() {
        light.turnOn();
    }

    public void undo() {
        light.turnOff();
    }
}

class LightOffCommand implements RemoteCommand {
    private Light light;

    LightOffCommand(Light light) {
        this.light = light;
    }

    public void execute() {
        light.turnOff();
    }

    public void undo() {
        light.turnOn();
    }
}

// Invoker — triggers commands and keeps a history for undo.
class RemoteControl {
    private Stack<RemoteCommand> history = new Stack<>();

    void pressButton(RemoteCommand command) {
        command.execute();
        history.push(command);
    }

    void pressUndo() {
        if (history.isEmpty()) {
            System.out.println("Nothing to undo.");
            return;
        }
        RemoteCommand last = history.pop();
        last.undo();
    }
}

public class Command {
    public static void main(String[] args) {
        Light light = new Light();
        RemoteControl remote = new RemoteControl();

        RemoteCommand onCommand = new LightOnCommand(light);
        RemoteCommand offCommand = new LightOffCommand(light);

        System.out.println("Press ON button:");
        remote.pressButton(onCommand);

        System.out.println("Press OFF button:");
        remote.pressButton(offCommand);

        System.out.println("Press UNDO button:");
        remote.pressUndo();

        System.out.println("Press UNDO button again:");
        remote.pressUndo();

        System.out.println("Press UNDO button once more (nothing left):");
        remote.pressUndo();
    }
}
